hal/tests/TESTS/pin_names/pinvalidate.py

```diff
@@ -46,286 +79,392 @@ def find_target(target_name = ''):
             realval = val
             depth = 0
-            while not re.match("(0x[0-9a-fA-F]+|[1-9][0-9]*|0[1-7][0-7]+|0b[01]+)[uUlL]{0,2}", realval):
+            while not re.match(
+                "(0x[0-9a-fA-F]+|[1-9][0-9]*|0[1-7][0-7]+|0b[01]+)[uUlL]{0,2}",
+                realval,
+            ):
                 try:
-                    realval = pinName_dict[realval]
+                    realval = pin_name_dict[realval]
                     depth += 1
-                except:
+                except KeyError:
                     break
                 if depth > 10:
                     break
 
             if realval in used_pins:
-                message = "already assigned to " + used_pins_friendly[used_pins.index(realval)]
-                invalid_items.append({'key': key, 'val': val, 'message': message})
+                message = (
+                    "already assigned to "
+                    + used_pins_friendly[used_pins.index(realval)]
+                )
+                invalid_items.append(
+                    {"key": key, "val": val, "message": message}
+                )
                 continue
 
             used_pins.append(realval)
-            used_pins_friendly.append(key + ' = ' + val)
+            used_pins_friendly.append(key + " = " + val)
     return invalid_items
 
-def arduino_existence_check(pinName_dict):
+
+def arduino_existence_check(pin_name_dict):
     required_pins = [
-        'ARDUINO_UNO_A0',
-        'ARDUINO_UNO_A1',
-        'ARDUINO_UNO_A2',
-        'ARDUINO_UNO_A3',
-        'ARDUINO_UNO_A4',
-        'ARDUINO_UNO_A5',
-        'ARDUINO_UNO_D0',
-        'ARDUINO_UNO_D1',
-        'ARDUINO_UNO_D2',
-        'ARDUINO_UNO_D3',
-        'ARDUINO_UNO_D4',
-        'ARDUINO_UNO_D5',
-        'ARDUINO_UNO_D6',
-        'ARDUINO_UNO_D7',
-        'ARDUINO_UNO_D8',
-        'ARDUINO_UNO_D9',
-        'ARDUINO_UNO_D10',
-        'ARDUINO_UNO_D11',
-        'ARDUINO_UNO_D12',
-        'ARDUINO_UNO_D13',
-        'ARDUINO_UNO_D14',
-        'ARDUINO_UNO_D15'
+        "ARDUINO_UNO_A0",
+        "ARDUINO_UNO_A1",
+        "ARDUINO_UNO_A2",
+        "ARDUINO_UNO_A3",
+        "ARDUINO_UNO_A4",
+        "ARDUINO_UNO_A5",
+        "ARDUINO_UNO_D0",
+        "ARDUINO_UNO_D1",
+        "ARDUINO_UNO_D2",
+        "ARDUINO_UNO_D3",
+        "ARDUINO_UNO_D4",
+        "ARDUINO_UNO_D5",
+        "ARDUINO_UNO_D6",
+        "ARDUINO_UNO_D7",
+        "ARDUINO_UNO_D8",
+        "ARDUINO_UNO_D9",
+        "ARDUINO_UNO_D10",
+        "ARDUINO_UNO_D11",
+        "ARDUINO_UNO_D12",
+        "ARDUINO_UNO_D13",
+        "ARDUINO_UNO_D14",
+        "ARDUINO_UNO_D15",
     ]
 
     invalid_items = []
 
     for pin in required_pins:
-        if pin not in pinName_dict:
+        if pin not in pin_name_dict:
             message = pin + " not defined"
-            invalid_items.append({'key': pin, 'val': '', 'message': message})
-    
+            invalid_items.append({"key": pin, "val": "", "message": message})
+
```

Comment:
Since you are only iterating through the pins once, you should use a generator or two and a list comprehension:

```python
# At the top
from itertools import chain # I prefer this way, but I think the pythonic way is just to "import itertools"

# In this function
a_pins = (f"ARDUINO_UNO_A{i}" for i in range(6))
d_pins = (f"ARDUINO_UNO_D{i} for i in range(16))

return  [{"key": pin, "val": "", "message": pin + " not defined"}] for pin in chain(a_pins, d_pins) if not in pin_name_dict]
```

Alternatively, since this looks like a helper function already, you could simply make this return a generator of dictionaries. That would then look like this (or close to it):

```python
# At the top
from itertools import chain # I prefer this way, but I think the pythonic way is just to "import itertools"

# In this function
a_pins = (f"ARDUINO_UNO_A{i}" for i in range(6))
d_pins = (f"ARDUINO_UNO_D{i} for i in range(16))

yield from ({"key": pin, "val": "", "message": pin + " not defined"}] for pin in chain(a_pins, d_pins) if not in pin_name_dict)
```

The consuming function can still wrap the generator in a `list` or just consume each dictionary one at a time.

Yet one more item: if you are using dictionaries a lot, but with the same keys and not mutating their values after creation, a `namedtuple` which would be faster, as it avoids some of the overhead of a `dict`.

Reply:
Thanks, I will consider the `namedtuple`. For now I'd prefer to keep all test cases as functions.